File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/services/proxies/HeapProxy.java
package com.github.ol_loginov.heaplibweb.services.proxies;

import com.github.ol_loginov.heaplibweb.repository.heap.*;
import lombok.RequiredArgsConstructor;
import org.apache.commons.lang3.NotImplementedException;
import org.netbeans.lib.profiler.heap.*;

import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.Properties;

@RequiredArgsConstructor
public class HeapProxy implements Heap {
	private final HeapEntity entity;
	private final HeapRepositories heapRepositories;

	protected JavaClassRepository getJavaClasses() {
		return heapRepositories.getJavaClasses();
	}

	protected FieldRepository getFields() {
		return heapRepositories.getFields();
	}

	@Override
	public boolean isWriteable() {
		return false;
	}

	@Override
	public List<JavaClass> getAllClasses() {
		return getJavaClasses().streamAllByHeapId(entity.getId())
			.map(this::proxyJavaClass)
			.toList();
	}

	private JavaClass proxyJavaClass(JavaClassEntity e) {
		return new JavaClassProxy(e, heapRepositories);
	}

	@Override
	public JavaClass getJavaClassByID(long javaclassId) {
		return getJavaClasses().findById(new JavaClassEntity.PK(entity.getId(), javaclassId))
			.map(this::proxyJavaClass)
			.orElse(null);
	}

	@Override
	public JavaClass getJavaClassByName(String fqn) {
		return getJavaClasses().findByHeapIdAndName(entity.getId(), fqn)
			.map(this::proxyJavaClass)
			.orElse(null);
	}

	@Override
	public Collection<JavaClass> getJavaClassesByRegExp(String regexp) {
		return getJavaClasses().findAllByHeapIdAndNameMatches(entity.getId(), regexp)
			.map(this::proxyJavaClass)
			.toList();
	}

	@Override
	public Iterable<Instance> getAllInstances() {
		throw new UnsupportedOperationException();
	}

	@Override
	public Iterable<Instance> getAllInstances(long instanceID) {
		throw new UnsupportedOperationException();
	}

	@Override
	public List<Instance> getBiggestObjectsByRetainedSize(int number) {
		throw new UnsupportedOperationException();
	}

	@Override
	public GCRoot getGCRoot(Instance instance) {
		throw new UnsupportedOperationException();
	}

	@Override
	public Collection<GCRoot> getGCRoots() {
		throw new UnsupportedOperationException();
	}

	@Override
	public Instance getInstanceByID(long instanceId) {
		throw new UnsupportedOperationException();
	}


	@Override
	public Iterator<Instance> getAllInstancesIterator() {
		throw new UnsupportedOperationException();
	}

	@Override
	public HeapSummary getSummary() {
		throw new UnsupportedOperationException();
	}

	@Override
	public Properties getSystemProperties() {
		throw new UnsupportedOperationException();
	}

	@Override
	public boolean isRetainedSizeComputed() {
		throw new UnsupportedOperationException();
	}

	@Override
	public boolean isRetainedSizeByClassComputed() {
		throw new UnsupportedOperationException();
	}
}

File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/services/proxies/HeapProxy.kt
package com.github.ol_loginov.heaplibweb.services.proxies

import com.github.ol_loginov.heaplibweb.repository.heap.ClassEntity
import com.github.ol_loginov.heaplibweb.repository.heap.HeapRepositories
import org.netbeans.lib.profiler.heap.*
import java.util.*

class HeapProxy(
    private val heapRepositories: HeapRepositories
) : Heap {
    private fun proxyJavaClass(e: ClassEntity): JavaClass = JavaClassProxy(e, heapRepositories)

    override fun isWriteable(): Boolean = false

    override fun getAllClasses(): List<JavaClass> = heapRepositories
        .classes.streamAll()
        .map { proxyJavaClass(it) }
        .toList()


    override fun getJavaClassByID(javaClassId: Long): JavaClass? = heapRepositories
        .classes.findById(javaClassId)
        ?.let { proxyJavaClass(it) }

    override fun getJavaClassByName(fqn: String): JavaClass? = heapRepositories
        .classes.findByName(fqn)
        ?.let { proxyJavaClass(it) }

    override fun getJavaClassesByRegExp(regexp: String): Collection<JavaClass> = heapRepositories
        .classes.findAllByNameRegex(regexp)
        .map { proxyJavaClass(it) }
        .toList()

    override fun getAllInstances(): Iterable<Instance> = throw UnsupportedOperationException()
    override fun getAllInstances(instanceID: Long): Iterable<Instance> = throw UnsupportedOperationException()
    override fun getBiggestObjectsByRetainedSize(number: Int): List<Instance> = throw UnsupportedOperationException()
    override fun getGCRoot(instance: Instance): GCRoot = throw UnsupportedOperationException()
    override fun getGCRoots(): Collection<GCRoot> = throw UnsupportedOperationException()
    override fun getInstanceByID(instanceId: Long): Instance = throw UnsupportedOperationException()
    override fun getAllInstancesIterator(): Iterator<Instance> = throw UnsupportedOperationException()
    override fun getSummary(): HeapSummary = throw UnsupportedOperationException()
    override fun getSystemProperties(): Properties = throw UnsupportedOperationException()
    override fun isRetainedSizeComputed(): Boolean = throw UnsupportedOperationException()
    override fun isRetainedSizeByClassComputed(): Boolean = throw UnsupportedOperationException()
}